### PR TITLE
enable proxyProvider

### LINKED ISSUE #3767: Implement “proxyProvider” logic on paywall
Once the iframe has been added to the DOM (#3766) our existing proxyProvider should send messages to it (#3765) to achieve the following:
* Get the current network 
* Get the user account (so that the paywall can check whether the user has keys…)
* Perform key purchases

### PR DESCRIPTION
**Pull request by cellog**
# Description

This is the same as #4002, but trimmed down and applied to `unlock_provider_base` instead of `master`. Some of this can probably still be trimmed, but it is small enough to be readable now.

# Issues

Fixes #3767

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [X] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [X] My code follows the style guidelines of this project, including naming conventions
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [X] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: paywall/src/unlock.js/web3Proxy.ts [hunk 1]
```diff
@@ -42,22 +110,104 @@ export default function web3Proxy(
...
     (window.web3.currentProvider.sendAsync || window.web3.currentProvider.send)
+  let proxyAccount: null | string = null
+  let proxyNetwork: string | number
+  let accountIframeReady = false
+  let currentLocks: Locks
+  let canUseUserAccounts: boolean = !window.web3
+  let hasNativeWeb3Wallet = !!window.web3
+
+  const useUnlockAccount = () =>
+    !hasNativeWeb3Wallet && proxyAccount && canUseUserAccounts
```
Comment (cnasc) [inline paywall/src/unlock.js/web3Proxy.ts]:
```suggestion
    proxyAccount && canUseUserAccounts
```

Comment (cellog) [reply paywall/src/unlock.js/web3Proxy.ts]: this was the original code, but it turns out it does not work. There are some cases where we have a native wallet present and it should use that (for instance, when we have not yet enabled access to the page)

File: paywall/src/messageTypes.ts [hunk 2]
```diff
@@ -125,6 +128,14 @@ export type Message =
       type: PostMessages.INITIATED_TRANSACTION
       payload: undefined
     }
+  | {
+      type: PostMessages.SHOW_ACCOUNTS_MODAL
```
Comment (cnasc) [inline paywall/src/messageTypes.ts]: I think we should consistently use `ACCOUNT` or `ACCOUNTS` on these names.

File: paywall/src/unlock.js/web3Proxy.ts [hunk 3]
```diff
@@ -42,22 +110,104 @@ export default function web3Proxy(
...
     postMessage: PostMessageToIframe<MessageTypes>
   ) => {
     // we don't have web3
-    hasWeb3 = false
+    // wait for locks to be retrieved
+    await waitFor(() => currentLocks)
+    // wait for the account iframe, then respond
+    await waitFor(() => accountIframeReady)
+    // we will use the proxy account!
+    if (!canUseUserAccounts) {
```
Comment (cnasc) [inline paywall/src/unlock.js/web3Proxy.ts]: There are a few places in this file where a variable is inverted, even though we already have another variable that represents the inversion of that variable. Is there another way to handle this?

File: paywall/src/unlock.js/web3Proxy.ts (hunk 1 above)
Comment (cnasc) [reply paywall/src/unlock.js/web3Proxy.ts]: Can you provide some more context? What are the cases where `canUseUserAccounts` and `!hasNativeWeb3Wallet` are different?

**Pull request merged**